witness.md

```diff
@@ -0,0 +1,599 @@
+
+* **Step 5**. No more rules are applicable, the witness contains only one
+    element, the execution ends successfully.
+
+
+## End Criteria
+
+The execution ends when there are no substitution rules applicable for this
+witness.
+
+### Building a single trie from the witness
+
+If we are building a single trie from the witness, then the only SUCCESS
+executon is when the following are true:
+
+- The execution state MUST match the End Criteria
+- There MUST be only one item left in the witness
+- This item MUST be of the type `Node`
+    
+In that case, this last item will be the root of the built trie.
+
+Every other end state is considered a FAILURE.
+
+
+### Building a Forest 
+
+We also can build a forest of tries with this approach, by adding a new
+Instruction `NEW_TRIE` and adjusting the success criteria a bit:
+
+- The execution state MUST match the End Criteria;
+- The items that are left in the witness MUST follow this pattern: `(Node
+    NEW_TRIE ... Node)`
+- Each `Node` element will be a root of a trie.
+
+Every other end state is considered a FAILURE.
+
+
+## Instructions & Parameters
+
+A single instruction consists of substitution rules and parameters.
+
+Each instruction MAY have one or more parameters.
+The parameters values MUST be encoded in the witness.
+
+That makes it different from the helper function parameters that MAY come from the stack or MAY come from the witness.
+
+
+## Substitution rules
+
+A substitution rule consists of 3 parts: 
+
+`[GUARD] PATTERN |=> RESULT`
+
+- to the left of the `|=>` sign:
+
+    - optional `GUARD` statements;
+
+    - the pattern to match against;
+
+- result, to the right of the `|=>` sign.
+
+
+### `GUARD`s
+
+Each substitution rule can have zero, one or multiple `GUARD` statements.
+Each `GUARD` statement looks like this:
+
+```
+GUARD <CONDITION>
+```
+
+For a substitution rule to be applicable, the `<CONDITION>` in its `GUARD` statement MUST be true.
+
+If a substitution rule has multiple `GUARD` statements, all of them MUST BE satisfied.
+
+If there are no `GUARD` statements, the substitution rule's applicability is
+only defined by the PATTERN.
+
+Example:
+```
+ GUARD NBITSET(mask) == 2
+|---- GUARD STATEMENT ---|
+
+ Node(n0) Node(n1) BRANCH(mask) |=> 
+ BranchNode{MAKE_VALUES_ARRAY(mask, n0, n1)}
+```
+
+For the example rule to be applicable both facts MUST be true:
+
+1. `mask` contains only 2 bits set to 1 (the rest are set to 0);
+
+2. to the left of `BRANCH` instruction there is at least 2 `Node`s.
+
+Fact (1) comes from the `GUARD` statement.
+
+
+### PATTERN
+
+`[NodeType(boundVar1)... NodeType(boundVarN)] Instruction[(param1... paramN)]`
+
+A pattern MUST contain a single instruction.
+A pattern MAY contain one or more `Node`s to the left of the instruction to
+match.
+An instruction MAY have one or more parameters.
+
+Pattern matching is happening by the types. `Node` type means any node is
```

Comment:
Concerns about unique readability of `NodeType` and `Instruction`. Should give a list of possible ones so that we see that there is no parsing ambiguity.